src/lib/db/feature-toggle-store.ts

```diff
@@ -171,34 +173,22 @@ export default class FeatureToggleStore implements IFeatureToggleStore {
 
     async setLastSeen(data: LastSeenInput[]): Promise<void> {
         const now = new Date();
-        const environmentArrays = this.mapMetricDataToEnvBuckets(data);
-        try {
-            for (const env of Object.keys(environmentArrays)) {
-                const toggleNames = environmentArrays[env].sort();
-                await this.db(FEATURE_ENVIRONMENTS_TABLE)
-                    .update({ last_seen_at: now })
-                    .where('environment', env)
-                    .whereIn(
-                        'feature_name',
-                        this.db(FEATURE_ENVIRONMENTS_TABLE)
-                            .select('feature_name')
-                            .whereIn('feature_name', toggleNames)
-                            .forUpdate()
-                            .skipLocked(),
-                    );
-
-                // Updating the toggle's last_seen_at also for backwards compatibility
-                await this.db(TABLE)
-                    .update({ last_seen_at: now })
-                    .whereIn(
-                        'name',
-                        this.db(TABLE)
-                            .select('name')
-                            .whereIn('name', toggleNames)
-                            .forUpdate()
-                            .skipLocked(),
-                    );
+        const enhancedData = data.map((value) => ({
+            feature_name: value.featureName,
+            last_seen_at: now,
+            environment: value.environment,
+        }));
+        const dataToPersist: typeof enhancedData = [];
+        for (const input of enhancedData) {
+            if (await this.exists(input.feature_name)) {
+                dataToPersist.push(input);
             }
+        }
+        try {
+            await this.db(FEATURE_ENVIRONMENTS_METRICS_TABLE)
+                .insert(dataToPersist)
+                .onConflict(['feature_name', 'environment'])
+                .merge(['last_seen_at']);
```

Comment:
essentially upsert - could not find a way to satisfy the constraint that the `name` exists in `features` table.  Because I am using onConflict to merge the last_seen_at column with the new value, I added the `EXISTS` check above.  The logic(assumption) is that the EXISTS operation in one of the least expensive operations.

Reply:
But you are doing exist for each datapoint, which means 10s or 100s of requests and roundtrips. Doesn't the onconflict/merge combo work without doing exists?

Reply:
It works for the conflict on the primary key - but not on the constraint to feature name of the features table.

Reply:
So if we try to persist metrics on a feature that does not exist  - the query fails

Reply:
I can try to bake the ON CONFLICT ON CONSTRAINT DO NOTHING to the table - if that works it would eliminate the need on all the EXISTS queries.

Reply:
why not skip on conflict?